lib/cadet/jobs/autograder/grading_job.ex

```diff
@@ -0,0 +1,197 @@
+defmodule Cadet.Autograder.GradingJob do
+  @moduledoc """
+  This module contains logic for finding answers to be graded and
+  processing/dispatching them as appropriate
+  """
+  use Cadet, :context
+
+  import Ecto.Query
+
+  alias Cadet.Assessments.{Answer, Assessment, Question, Submission}
+  alias Cadet.Autograder.Utilities
+
+  def grade_all_due_yesterday do
+    for assessment <- Utilities.fetch_assessments_due_yesterday() do
+      assessment.id
+      |> Utilities.fetch_submissions()
+      |> Enum.map(fn %{student_id: student_id, submission: submission} ->
+        if submission do
+          update_submission_status_to_submitted(submission)
+        else
+          insert_empty_submission(%{student_id: student_id, assessment: assessment})
+        end
+      end)
+      |> Enum.each(fn submission ->
+        grade_individual_submission(submission, assessment)
+      end)
+    end
+  end
+
+  @doc """
+  Exposed as public function in case future mix tasks are needed to regrade
+  certain submissions. Manual grading can also be triggered from iex with this
+  function.
+
+  Takes in submission to be graded. Submission will be graded regardless of its
+  assessment's close_by date or submission status.
+
+  Every answer will be regraded regardless of its current autograding status.
+  """
+  def force_grade_individual_submission(submission = %Submission{}) do
+    assessment =
+      if Ecto.assoc_loaded?(submission.assessment) do
+        submission.assessment
+      else
+        submission |> Repo.preload(:assessment) |> Map.get(:assessment)
+      end
+
+    assessment = preprocess_assessment_for_grading(assessment)
+    grade_individual_submission(submission, assessment, true)
+  end
+
+  # This function requires that assessment questions are already preloaded in sorted
+  # order for autograding to function correctly.
+  defp grade_individual_submission(
+         %Submission{id: submission_id},
+         %Assessment{questions: questions},
+         regrade \\ false
+       ) do
+    answers =
+      Answer
+      |> where(submission_id: ^submission_id)
+      |> order_by(:question_id)
+      |> Repo.all()
+
+    grade_submission_question_answer_lists(
+      submission_id,
+      questions,
+      answers,
+      regrade
+    )
+  end
+
+  defp preprocess_assessment_for_grading(assessment = %Assessment{}) do
+    if Ecto.assoc_loaded?(assessment.questions) do
```

Comment:
Now I'm thinking again, would it ever actually be loaded since this function is called from `force_grade_individual_submission/1` and `Assessments.finalise_submission/2` both of which don't preload questions.

Reply:
TBH I'm not sure how this will actually be used. I wanted to have it there in case all submissions for an assessment needed to be regraded, in which case this makes it easy to add a `force_regrade_assessment` that calls `preprocess_assessment_for_grading` only once instead of reloading the same questions 500 times.

Reply:
Fair enough.